pkg/reconciler/v1alpha1/serverlessservice/serverlessservice.go

```diff
@@ -0,0 +1,346 @@
+	if err != nil {
+		logger.Error(fmt.Sprintf("Error obtaining private service endpoints: %s", psn), zap.Error(err))
+		return err
+	}
+	logger.Debugf("Public endpoints: %s", spew.Sprint(srcEps))
+
+	sn := names.PublicService(sks)
+	eps, err := r.endpointsLister.Endpoints(sks.Namespace).Get(sn)
+
+	if errors.IsNotFound(err) {
+		logger.Infof("K8s endpoints %q does not exist; creating.", sn)
+		sks.Status.MarkEndpointsUnknown("Creating")
+		eps, err = r.KubeClientSet.CoreV1().Endpoints(sks.Namespace).Create(resources.MakePublicEndpoints(sks, srcEps))
+		if err != nil {
+			logger.Errorw(fmt.Sprintf("Error creating K8s Endpoints %s: ", sn), zap.Error(err))
+			return err
+		}
+		logger.Infof("Created K8s Endpoints: %q", sn)
+	} else if err != nil {
+		logger.Errorw(fmt.Sprintf("Error getting K8s Endpoints %s: ", sn), zap.Error(err))
+		return err
+	} else if !metav1.IsControlledBy(eps, sks) {
+		sks.Status.MarkEndpointsNotOwned("Endpoints", sn)
+		return fmt.Errorf("SKS: %q does not own Endpoints: %q", sks.Name, sn)
+	} else {
+		want := eps.DeepCopy()
+		want.Subsets = srcEps.Subsets
+
+		if !equality.Semantic.DeepEqual(want.Subsets, eps.Subsets) {
+			log.Infof("Public K8s Endpoints changed; reconciling: %s", sn)
+			if _, err = r.KubeClientSet.CoreV1().Endpoints(sks.Namespace).Update(want); err != nil {
+				logger.Errorw(fmt.Sprintf("Error updating public K8s Endpoints %s: ", sn), zap.Error(err))
+				return err
+			}
+		}
+	}
+	if hasEndpoints(eps) {
+		sks.Status.MarkEndpointsPopulated()
+		sks.Status.ServiceName = sn
+	} else {
+		sks.Status.MarkEndpointsUnknown("NoHealthyBackends")
+	}
+	logger.Debugf("Done reconciling public K8s endpoints %s", sn)
+	return nil
+}
+
+// hasEndpoints returns true if Endpoints resource has at least one endpoint.
+func hasEndpoints(eps *corev1.Endpoints) bool {
+	for _, ss := range eps.Subsets {
+		if len(ss.Addresses) > 0 {
+			return true
+		}
+
+	}
+	return false
+}
+
+func (r *reconciler) reconcilePrivateService(ctx context.Context, sks *netv1alpha1.ServerlessService) error {
+	logger := logging.FromContext(ctx)
+	sn := names.PrivateService(sks)
+	svc, err := r.serviceLister.Services(sks.Namespace).Get(sn)
+	if errors.IsNotFound(err) {
+		logger.Infof("K8s service %q does not exist; creating.", sn)
+		sks.Status.MarkEndpointsUnknown("Creating")
+		svc = resources.MakePrivateService(sks)
+		_, err := r.KubeClientSet.CoreV1().Services(sks.Namespace).Create(svc)
+		if err != nil {
+			logger.Errorw(fmt.Sprintf("Error creating K8s Service %s: ", sn), zap.Error(err))
+			return err
+		}
+		logger.Infof("Created K8s service: %q", sn)
+	} else if err != nil {
+		logger.Errorw(fmt.Sprintf("Error getting K8s Service %s: ", sn), zap.Error(err))
+		return err
+	} else if !metav1.IsControlledBy(svc, sks) {
+		sks.Status.MarkEndpointsNotOwned("Service", sn)
+		return fmt.Errorf("SKS: %q does not own Service: %q", sks.Name, sn)
+	} else {
+		tmpl := resources.MakePrivateService(sks)
```

Comment:
A function would make these 4 lines more clear. It looks like you are trying to preserve fields not touched by private service reconciliation?

Reply:
Yep, we own only part of the spec (selector and portspec). So just set those, since there's lots of stuff that K8s manages itself.
For now I'll add a comment, since the template is different in the two cases, where this could benefit from a function extraction.